### PR TITLE
Actionspace

### PR DESCRIPTION
**Pull request by Leuxll**
Changelog:

Changes pertaining to the refactoring of ActionSpace Class will go down the file structure outlining the changes made to the ActionSpace feature and any architectural changes/modifications chosen.

### Entities
**ActionSpace2**
- Added ActionSpace2, this would be the new ActionSpace Entities class that has no logic and only contains data, unlike the previous implementation of the ActionSpace

**Things that need to be done within this class**
- [ ] ActionSpace extends from an abstract class Cell that has performAction as a method. This right now returns a string, but the thing is this is called in outer layers. Changed this by putting the perform action as an interactor implementation and the use case interface would be used as the outward output boundary.

**Card**
- Changed ActionSpace so that it becomes a list of Objects that only pertains to that type of action space, in the past every ActionSpace had all three decks of cards.

### Interactors
**ActionSpaceCreationInteractor**
- Handles the Creation of ActionSpace and implements the ActionSpaceCreationUseCase, which is what is going to be called.

**CardMapperInteractor**
- Maps the data from the input file to the correct data structure that is going to go into the Entities themselves.

**PerformActionSpaceCardInteractor**
- This contains all of the old implementations of performAction and also the generateRandomCard. This also implements an interface PerformActionSpaceUseCase that should be called when performAction is needed.

###Persistence Layer
- added this layer as this should be the layer that handles all of the loadings of cards and accessing of data when loading in the cards.

**TextFileTranslator**
- Added loading of the cards from the txt file

### UsesCases
- Theses contains all of the use cases in which should my interfaces be implemented by the interactions.

Comment on changes that you want to be done and the next time we meet, we will need to discuss when the other classes call the ActionSpace entity.

### CODE REVIEW

File: src/main/java/Entities/Player.java [hunk 1]
```diff
@@ -171,112 +132,69 @@ public String steal(Player victim) {
...
 
-    /**
-     * Method to move the player by x number of steps
-     * @param step the steps to move the player to
-     */
+
     public void move(int step) {
         position += step;
         if (position > 39) {
-            //Money on passing go
```
Comment (Cindywwang) [inline src/main/java/Entities/Player.java:181]: Thanks for catching this.

File: src/main/java/Entities/ActionSpace2.java [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+        return cards;
+    }
+
+    @Override
+    public String performAction(Player player, Board board) {
+        // TODO Auto-generated method stub
+        return null;
+    }
+
+}
```
Comment (harshkpatel) [inline src/main/java/Entities/ActionSpace2.java:31]: I think a better naming convention can be used for this class, and on top of that change the name of the other action space class with respect to this one.

File: src/main/java/Entities/Card.java [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+    public String getActionType() {
+        return this.actiontype;
+    }
+
+    public Integer getAmount() {
+        return this.amount;
+    }
+
+}
```
Comment (harshkpatel) [inline src/main/java/Entities/Card.java:33]: Good choice of not including any logic in any of the entities.

File: src/main/java/Entities/Player.java [hunk 4]
```diff
@@ -195,88 +212,96 @@ public String riggedRoll(int rig) {
...
+    /**
+     * Sets the current players properties to another ArrayList of properties
+     * @param properties the ArrayList of properties to change it to
+     */
+    public void setProperties(ArrayList<Property> properties) {
+        this.properties = properties;
+    }
+
+
 }
```
Comment (harshkpatel) [inline src/main/java/Entities/Player.java:307]: Good job of removing all the logic out of this entity, some of the JavaDoc needs to be written better, i.e. remove any time you say "basically" and any other unprofessional words.

File: src/main/java/Interactors/ActionSpaceCreationInteractor.java [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+     * 
+     * @return the action space
+     */
+    public ActionSpace2 loadCards() throws IOException {
+        // TODO Auto-generated method stub
+        HashMap<String, List<Card>> cards = cardAccess.CardMapperJailCards(loadCards.loadCards());
+        return new ActionSpace2(cards);
+    }
+
+}
```
Comment (harshkpatel) [inline src/main/java/Interactors/ActionSpaceCreationInteractor.java:42]: good choice of adding an interactor to add another layer, to remove dependency between the entities and the rest of the code.

File: src/main/java/Interactors/CardMapperInteractor.java [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+                Integer amount = Integer.parseInt(arr[3]);
+                Card card = new Card(type, action, actiontype, amount);
+                cards.add(card);
+            }
+        }
+        comChestCards.put("comchest", cards);
+        return comChestCards;
+    }
+
+}
```
Comment (harshkpatel) [inline src/main/java/Interactors/CardMapperInteractor.java:82]: Good choice of moving the logic outward.

File: src/main/java/Persistence/TextFileTranslator.java [hunk 7]
```diff
@@ -207,4 +209,15 @@ public ArrayList<String[]> loadProperties() throws FileNotFoundException {
...
+     * 
+     * @return a list of strings that represents all of the cards in the game
+     * @throws IOException
+     *
+     */
+    public List<String> loadCards() throws IOException {
+
+        return Files.readAllLines(Paths.get("src/save/cards.txt"));
+
+    }
```
Comment (harshkpatel) [inline src/main/java/Persistence/TextFileTranslator.java:222]: Good choice of making the persistence a separate layer

**Review by harshkpatel — APPROVED** [review]
Overall I do not have too many comments about the refactor, as you do get rid of a lot of code smells and actually implement clean architecture. One big point is the naming conventions used, alot of them are poorly written/lack professionalism; however I am going to leave this for a fix in a future PR, where in addition to the changes of naming conventions, the changes in action space are reflected in the rest of the code. I am certain that the changes you made right now are not reflected in the rest of the codebase and that the game will not actually run because the rest of the entities need to be refactored. For this reason that the refactoring can be done in a later PR to accommodate for all the other changes.

**Review by harshkpatel — COMMENTED** [review]
No additional comments, besides good changes to make the code more scalable.

**Review by kRichard32 — COMMENTED** [review]
Overall, good design of separating the use cases from the entities and having logic being handled elsewhere. 
However, with this new implementation, we need to fix the other classes, specifically, the game creation interactor, the main tree handler and the property class, as the objects are being created differently and that the main tree handler no longer calls performAction.

**Review by kRichard32 — APPROVED** [review]

**Review by benxu3 — APPROVED** [review]
Good work refactoring the action space entity to comply with clean architecture principles. Remember to remove ActionSpace.java after we merge in the PR with the Property and Corner Tiles PR.

**Pull request merged**